l5kit/l5kit/simulation/unroll.py

```diff
@@ -0,0 +1,226 @@
+                                        self.sim_cfg.distance_th_close)
+
+        if self.sim_cfg.num_simulation_steps is None:
+            range_unroll = range(self.sim_cfg.start_frame_index, len(sim_dataset))
+        else:
+            end_frame_index = min(len(sim_dataset), self.sim_cfg.num_simulation_steps + self.sim_cfg.start_frame_index)
+            range_unroll = range(self.sim_cfg.start_frame_index, end_frame_index)
+
+        for frame_index in tqdm(range_unroll):
+            next_frame_index = frame_index + 1
+            should_update = next_frame_index != range_unroll.stop
+
+            # AGENTS
+            if not self.sim_cfg.use_agents_gt:
+                agents_input = sim_dataset.rasterise_agents_frame_batch(frame_index)
+                agents_input_dict = default_collate(list(agents_input.values()))
+                agents_output_dict = self.model_agents(agents_input_dict)
+                if should_update:
+                    self.update_agents(sim_dataset, next_frame_index, agents_input_dict, agents_output_dict)
+
+            # EGO
+            if not self.sim_cfg.use_ego_gt:
+                ego_input = sim_dataset.rasterise_frame_batch(frame_index)
+                ego_input_dict = default_collate(ego_input)
+                ego_output_dict = self.model_ego(ego_input_dict)
+                if should_update:
+                    self.update_ego(sim_dataset, next_frame_index, ego_input_dict, ego_output_dict)
+
+        simulated_outputs: List[SimulationOutputs] = []
+        for scene_idx in scene_indices:
+            simulated_outputs.append(SimulationOutputs(scene_idx, sim_dataset))
+        return simulated_outputs
+
+    @staticmethod
+    def update_agents(dataset: SimulationDataset, frame_idx: int, input_dict: Dict[str, torch.Tensor],
+                      output_dict: Dict[str, torch.Tensor]) -> None:
+        """Update the agents in frame_idx (across scenes) using agents_output_dict
+
+        :param dataset: the simulation dataset
+        :param frame_idx: index of the frame to modify
+        :param input_dict: the input to the agent model
+        :param output_dict: the output of the agent model
+        :return:
+        """
+
+        agents_update_dict: Dict[Tuple[int, int]: np.ndarray] = {}
+        input_dict = {k: v.cpu().numpy() for k, v in input_dict.items()}
+        output_dict = {k: v.cpu().numpy() for k, v in output_dict.items()}
+
+        world_from_agent = input_dict["world_from_agent"]
+        yaw = input_dict["yaw"]
+        pred_trs = transform_points(output_dict["positions"][:, :1], world_from_agent)[:, 0]
+        pred_yaws = yaw + output_dict["yaws"][:, 0]
+
+        next_agents = np.zeros(len(yaw), dtype=AGENT_DTYPE)
+        next_agents["centroid"] = pred_trs
+        next_agents["yaw"] = pred_yaws
+        next_agents["track_id"] = input_dict["track_id"]
+        next_agents["extent"] = input_dict["extent"]
+        next_agents["label_probabilities"][:, 3] = 1
```

Comment:
I think we should create some enums for the fields here, I don't really know what is the `3` for instance.

Reply:
I can actually get the index of the label here. We should definitely add some enum later though